### PR TITLE
IGNITE-16294 .NET: Fix System.Enum serialization

### PR DESCRIPTION
**Pull request by Vladsz83**
.NET binary writer doesn't get `System.Enum` field as an enum. Happens because `System.Enum::IsEnum == false` and because we build user-type serialization based on declared field types (`FieldInfo::FieldType`), not on types of real assigned values.

Solution: Fixed the type caster so that it deals with `System.Enum`.

### CODE REVIEW

**Review by nizhikov — APPROVED** [review]

**Comment by nizhikov** [comment]
LGTM 

@ptupitsyn Can you, please, take a look at this changes, also.

File: modules/platforms/dotnet/Apache.Ignite.Core/Impl/Common/TypeCaster.cs [hunk 1]
```diff
@@ -87,9 +89,39 @@ private static class Casters<TFrom>
                 }
 
                 var paramExpr = Expression.Parameter(typeof(TFrom));
-                var convertExpr = Expression.Convert(paramExpr, typeof(T));
 
-                return Expression.Lambda<Func<TFrom, T>>(convertExpr, paramExpr).Compile();
+                if (typeof(TFrom) == typeof(Enum))
+                    return Expression.Lambda<Func<TFrom, T>>(TryConvertRawEnum(typeof(T), paramExpr),
```
Comment (ptupitsyn) [inline modules/platforms/dotnet/Apache.Ignite.Core/Impl/Common/TypeCaster.cs:94]: What if `TryConvertRawEnum` returns `null`?

**Review by ptupitsyn — CHANGES_REQUESTED** [review]

File: modules/platforms/dotnet/Apache.Ignite.Core/Impl/Common/TypeCaster.cs (hunk 1 above)
Comment (Vladsz83) [reply modules/platforms/dotnet/Apache.Ignite.Core/Impl/Common/TypeCaster.cs:94]: Would fail with same conversion error.

Comment (ptupitsyn) [reply modules/platforms/dotnet/Apache.Ignite.Core/Impl/Common/TypeCaster.cs:94]: `Expression.Lambda()` will throw `ArgumentNullException` if `TryConvertRawEnum` returns null. `ArgumentException` is always a bug. Let's check for null and throw `InvalidCastException` explaining that `Enum` can't be converted to the specified type.

**Review by ptupitsyn — APPROVED** [review]

**Comment by ptupitsyn** [comment]
Looks good to me.

**Pull request merged**